### PR TITLE
Feature lwdev 6980 monthly operations report

### PR DESCRIPTION
**Pull request by GSailen**

### CODE REVIEW

File: src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        {
+            return string.IsNullOrWhiteSpace(_tradingReportFilter.LegalEntityRegex)
+                   || (!string.IsNullOrWhiteSpace(legalEntity)
+                       && Regex.IsMatch(legalEntity, _tradingReportFilter.LegalEntityRegex, RegexOptions.IgnoreCase));
+        }
+
+        public async Task PerformReporting()
+        {
+            var from = _systemClock.UtcNow.Date.AddMonths(-1);
+            var to = _systemClock.UtcNow.Date;
```
Comment (andrey-anokhin) [inline src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs:75]: I thought is should be report for previous calendar month, not report for last 30 days.

File: src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            //gather data concurrently, await & filter & convert
+            var pendingPositionsTask = _tradeMonitoringReadingApi.PendingOrders();
+            var accountsTask = _accountsApi.GetAllAccounts();
+
+            var accounts = (await accountsTask).Where(x => LegalEntityPredicate(x.LegalEntity))
+                .Select(x => _convertService.Convert<DataReaderAccountBackendContract, Account>(x))
+                .ToList();
+            var clientIds = accounts.Select(x => x.ClientId).Distinct().ToArray();
+            var accountIds = accounts.Select(x => x.Id).Distinct().ToHashSet();
+            var accountClients = accounts.ToDictionary(x => x.Id, x => x.ClientId);
```
Comment (andrey-anokhin) [inline src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs]: accountClients.Keys will contain the same data like accountIds.
accountClients.ContainKey can be used

File: src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            if (failedNotifications.Any())
+            {
+                //TODO handle failed notifications
+                await _log.WriteWarningAsync(nameof(NotificationGenerator), nameof(TradingReportService),
+                    nameof(PerformReporting), $"{failedNotifications.Count} notifications failed to be send.",
+                    anyException, _systemClock.UtcNow.DateTime);
+            }
+        }
+
+        private static MonthlyTradingNotification PrepareNotification(int reportMonth,
```
Comment (andrey-anokhin) [inline src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs:178]: Very strange manipulations with collections... Let's discuss it.

File: src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            IReadOnlyList<AccountHistory> accountTransactions)
+        {
+            var filteredAccounts = accounts.Where(x => x.ClientId == clientId)
+                .Select(x =>
+                {
+                    x.AccountTransactions = accountTransactions.Where(at => at.ClientId == clientId
+                                                                            && at.AccountId == x.Id)
+                        .OrderByDescending(at => at.Date).ToList();
+                    var firstTransaction = x.AccountTransactions.LastOrDefault();
+                    x.InitialBalance = (firstTransaction?.Balance - firstTransaction?.Amount) ?? x.Balance;
```
Comment (andrey-anokhin) [inline src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs]: This is not completely correct.

**Review by andrey-anokhin — CHANGES_REQUESTED** [review]
Binaries should be removed from repo

File: src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs (hunk 2 above)
Comment (GSailen) [reply src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs]: Fixed

File: src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs (hunk 1 above)
Comment (GSailen) [reply src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs:75]: Not an issue

File: src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs (hunk 3 above)
Comment (GSailen) [reply src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs:178]: Might be optimised for sure. But it's ok for the particular case.

File: src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs (hunk 4 above)
Comment (GSailen) [reply src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs]: InitialBalance removed.

**Pull request merged**